src/domain/helpers/ExcelBuilder.ts

```diff
@@ -353,6 +383,56 @@ export class ExcelBuilder {
         }
     }
 
+    private buildTeiEventsForCustomTemplates(options: {
+        template: Template;
+        dataSource: TrackerEventRowDataSource;
+        payload: TrackerProgramPackage;
+    }): DataPackageData[] {
+        const { template, dataSource, payload } = options;
+
+        if (template.type !== "custom") return payload.dataEntries;
+        if (!dataSource.onlyLastEvent) return payload.dataEntries;
+
+        const eventsByTei = _(payload.dataEntries)
+            .sortBy(dataSource.sortBy ?? "")
+            .groupBy(event => event.trackedEntityInstance)
+            .value();
+
+        return _(eventsByTei)
+            .values()
+            .map(event => {
+                const sorted = _(event).sortBy("id").value();
+                return _(sorted).last();
+            })
+            .compact()
+            .value();
+    }
+
+    private buildTeisForCustomTemplates(options: {
+        template: Template;
+        dataSource: TeiRowDataSource | TrackerEventRowDataSource;
+        payload: TrackerProgramPackage;
+    }): TrackedEntityInstance[] {
+        const { template, dataSource, payload } = options;
+
+        if (template.type !== "custom") return payload.trackedEntityInstances;
+        if (dataSource.type === "rowTei" && !dataSource.skipTeisWithoutEvents) return payload.trackedEntityInstances;
+
+        const eventsByTei = _(payload.dataEntries)
+            .groupBy(dataEntry => dataEntry.trackedEntityInstance)
+            .value();
+
+        const dataEntriesTeisNoEvents = _(payload.trackedEntityInstances)
+            .filter(tei => {
+                const events = eventsByTei[tei.id] || [];
+                return events.length > 0;
```

Comment:
Variable is `dataEntriesTeisNoEvents` but check is "if it has events", am i missing something?